[AMDGPU] Add MaxMemoryClauseSchedStrategy

ruiling opened the pull request:
The major behaviors of the max memory clause schedule strategy includes:
1. Try to cluster memory instructions more aggressively.
2. Try to schedule long latency load earlier than short latency
   instruction.

I tested locally against about 470 real shaders and got the perf changes:
(only count perf changes over +/-10%):
About 15 shaders improved 10%~40%.
Only 3 shaders drops ~10%.

(This was tested together with another change which increases the maximum clustered dword from 8 to 32).
I will make another change to make that threshold configurable.

ruiling committed:
[AMDGPU] Add a test for grouping image instructions

This is a motiviting example that drives us to do better on grouping
image sample instructions.

ruiling committed:
[AMDGPU] Add AMDGPU specific tryCandidate()

The AMDGPU specific version mainly includes two major differences:
1. Try to cluster memory instructions more aggressively.
2. Try to schedule long latency load earlier than short latency
   instruction.

I tested locally against about 470 real shaders and got the perf changes:
(only count perf changes over +/-10%):
About 15 shaders improved 10%~40%.
Only 3 shaders drops ~10%.

(This was tested together with another change which increases the maximum clustered dword from 8 to 32).
I will make another change to make that threshold configurable.

arsenm commented on llvm/lib/Target/AMDGPU/GCNSchedStrategy.cpp:
```diff
@@ -63,6 +63,10 @@ static cl::opt<bool> GCNTrackers(
     cl::init(false));
 
+static cl::opt<bool> UseAMDGPUScheduleHeuristic(
+    "amdgpu-use-amdgpu-schedule-heuristic", cl::Hidden,
```
Name and description too general

Also it should be on by default, if an option at all

ruiling replied:
I think it is OK to be default on (the only difference would be lots of test changes). But I would like we could keep it off for the first submission (at least before we see some performance results on compute benchmarks). We may still need fine-tuning the heuristics after more extensive testing. We can turn it on later once this version's tryCandidate() becomes somewhat stable and do not regress much real world cases. I am not sure how such kind of performance testing is done in compute side. It would better we can get some performance numbers for the change. Please let me know if you still have concern on this.


> Name and description too general

Do you have any suggestion? Maybe `amdgpu-use-custom-try-candidate`?

kerbowa replied:
Some of the scheduling changes I'm working on involve having multiple selectable versions of tryCandidate. I think we can assume that future work on the scheduler will always include a custom tryCandidate override. We also already override it in the max-ilp scheduler. So I agree, the name is too broad. I think you should make this a separate SchedStrategy and have it be selectable that way.

ruiling replied:
Thanks @kerbowa for the suggestion! I have make it a separate strategy.

arsenm commented on llvm/test/CodeGen/AMDGPU/group-image-instructions.ll:
```diff
@@ -0,0 +1,488 @@
+; GFX11-NEXT:    exp mrt0 v0, v1, off, off done
+; GFX11-NEXT:    s_endpgm
+.entry:
+  %0 = call i64 @llvm.amdgcn.s.getpc()
```
Use named values in tests

ruiling committed:
Move to new MaxMemoryClause schedule strategy

ruiling committed:
Use named IR values

ruiling committed:
Also add a function attribute to control it

The graphics frontend could not set command-line option per function.
Add the function attribute to allow the frontend to set it per
function.

arsenm commented on llvm/lib/Target/AMDGPU/AMDGPUTargetMachine.cpp:
```diff
@@ -1290,6 +1312,11 @@ ScheduleDAGInstrs *GCNPassConfig::createMachineScheduler(
 
+  if (EnableMaxMemoryClauseSchedStrategy ||
+      C->MF->getFunction().hasFnAttribute(
+          "amdgpu-enable-max-memory-clause-scheduling-strategy"))
```
probably should avoid the word "enable", and this isn't very flexible in terms of selecting from the chain options we have here. How about an attribute with a named strategy?

ruiling replied:
The suggestion sounds pretty good. I have changed the attribute to named strategy.

ruiling committed:
Add amdgpu-sched-strategy for selecting sched strategy

ruiling commented:
ping

arsenm commented on llvm/lib/Target/AMDGPU/GCNSchedStrategy.cpp:
```diff
@@ -615,6 +615,137 @@ bool GCNMaxILPSchedStrategy::tryCandidate(SchedCandidate &Cand,
+    : GCNSchedStrategy(C) {
+  SchedStages.push_back(GCNSchedStageID::MemoryClauseInitialSchedule);
+}
+/// GCNMaxMemoryClauseSchedStrategy tries best to clause memory instructions as
```
Line separate

arsenm commented on llvm/lib/Target/AMDGPU/AMDGPUTargetMachine.cpp:
```diff
@@ -434,6 +434,12 @@ static cl::opt<bool> EnableMaxIlpSchedStrategy(
     cl::desc("Enable scheduling strategy to maximize ILP for a single wave."),
     cl::Hidden, cl::init(false));
 
+static cl::opt<bool> EnableMaxMemoryClauseSchedStrategy(
+    "amdgpu-enable-max-memory-clause-scheduling-strategy",
+    cl::desc("Enable scheduling strategy to maximize memory clause for a "
+             "single wave."),
+    cl::Hidden, cl::init(false));
```
If we're going to keep cl::opts for this, it should mirror the attribute. One enum cl::opt with the same options as the attribute

arsenm commented on llvm/lib/Target/AMDGPU/AMDGPUTargetMachine.cpp:
```diff
@@ -1287,9 +1309,18 @@ ScheduleDAGInstrs *GCNPassConfig::createMachineScheduler(
+  StringRef SchedStrategy =
+      SchedStrategyAttr.isValid() ? SchedStrategyAttr.getValueAsString() : "";
+
+  if (EnableMaxIlpSchedStrategy || SchedStrategy == "max-ilp")
```
Treat the option as an alternative source to the attribute?

ruiling committed:
Introduce a new commandline option: mdgpu-sched-strategy

arsenm commented on llvm/lib/Target/AMDGPU/GCNSchedStrategy.cpp:
```diff
@@ -615,6 +615,138 @@ bool GCNMaxILPSchedStrategy::tryCandidate(SchedCandidate &Cand,
+    return TryCand.Reason != NoCand;
+
+  // Avoid exceeding the target's limit.
+  if (DAG->isTrackingPressure() &&
+      tryPressure(TryCand.RPDelta.Excess, Cand.RPDelta.Excess, TryCand, Cand,
+                  RegExcess, TRI, DAG->MF))
+    return TryCand.Reason != NoCand;
+
+  // Avoid increasing the max critical pressure in the scheduled region.
+  if (DAG->isTrackingPressure() &&
+      tryPressure(TryCand.RPDelta.CriticalMax, Cand.RPDelta.CriticalMax,
+                  TryCand, Cand, RegCritical, TRI, DAG->MF))
+    return TryCand.Reason != NoCand;
```
Put these under one isTrackingPressure check

arsenm commented on llvm/lib/Target/AMDGPU/AMDGPUTargetMachine.cpp:
```diff
@@ -561,6 +561,18 @@ createGCNMaxILPMachineScheduler(MachineSchedContext *C) {
+  const GCNSubtarget &ST = C->MF->getSubtarget<GCNSubtarget>();
+  ScheduleDAGMILive *DAG = new GCNScheduleDAGMILive(
+      C, std::make_unique<GCNMaxMemoryClauseSchedStrategy>(C));
+  DAG->addMutation(createLoadClusterDAGMutation(DAG->TII, DAG->TRI));
+  if (ST.shouldClusterStores())
+    DAG->addMutation(createStoreClusterDAGMutation(DAG->TII, DAG->TRI));
```
Does this end up being redundant with the load and store clustering?

ruiling replied:
Sorry I don't get the point here, why it is redundant? with what? (This is mostly inherited from the createGCNMaxOccupancyMachineScheduler())

arsenm replied:
I mean if you cluster loads and stores, you're helping with clause formation

ruiling replied:
Yes, it's true.

arsenm commented on llvm/lib/Target/AMDGPU/GCNSchedStrategy.cpp:
```diff
@@ -615,6 +615,138 @@ bool GCNMaxILPSchedStrategy::tryCandidate(SchedCandidate &Cand,
+      return TryCand.Reason != NoCand;
+
+    // Fall through to original instruction order.
+    if ((Zone->isTop() && TryCand.SU->NodeNum < Cand.SU->NodeNum) ||
+        (!Zone->isTop() && TryCand.SU->NodeNum > Cand.SU->NodeNum)) {
```
```suggestion
    if ((Zone->isTop() == TryCand.SU->NodeNum < Cand.SU->NodeNum)) {
```

Assuming you don't need to worry about the NodeNum equal case

ruiling replied:
I am not sure whether we would have the equal case, but I think that is meaningless case. I just put an assert to catch that.

arsenm commented on llvm/test/CodeGen/AMDGPU/group-image-instructions.ll:
```diff
@@ -0,0 +1,488 @@
+; NOTE: Assertions have been autogenerated by utils/update_llc_test_checks.py UTC_ARGS: --version 5
+; RUN: llc -mtriple=amdgcn -mcpu=gfx1100 -verify-machineinstrs < %s | FileCheck -check-prefixes=GFX11 %s
```
Don't need -verify-machineinstrs (also use a full triple)

arsenm commented on llvm/test/CodeGen/AMDGPU/group-image-instructions.ll:
```diff
@@ -0,0 +1,488 @@
+declare float @llvm.amdgcn.interp.inreg.p2(float, float, float) #3
+declare <2 x i32> @llvm.amdgcn.s.buffer.load.v2i32(<4 x i32>, i32, i32 immarg) #8
+
+attributes #2 = { alwaysinline nounwind memory(readwrite) "amdgpu-sched-strategy"="max-memory-clause" "InitialPSInputAddr"="2" "amdgpu-color-export"="1" "amdgpu-depth-export"="0" "amdgpu-memory-bound"="false" "amdgpu-prealloc-sgpr-spill-vgprs" "amdgpu-unroll-threshold"="700" "amdgpu-wave-limiter"="false" "denormal-fp-math-f32"="preserve-sign" "target-features"=",+wavefrontsize64,-cumode" }
```
Remove unrelated attributes. Probably don't need target-features? (which also has a stray comma)

ruiling committed:
Refine code and test

arsenm commented on llvm/test/CodeGen/AMDGPU/group-image-instructions.ll:
```diff
@@ -1,11 +1,11 @@
 ; NOTE: Assertions have been autogenerated by utils/update_llc_test_checks.py UTC_ARGS: --version 5
-; RUN: llc -mtriple=amdgcn -mcpu=gfx1100 -verify-machineinstrs < %s | FileCheck -check-prefixes=GFX11 %s
+; RUN: llc -mtriple=amdgcn-amd-amdpa -mcpu=gfx1100 < %s | FileCheck -check-prefixes=GFX11 %s
```
```suggestion
; RUN: llc -mtriple=amdgcn-amd-amdpal -mcpu=gfx1100 < %s | FileCheck -check-prefixes=GFX11 %s
```
Typo

arsenm commented on llvm/lib/Target/AMDGPU/GCNSchedStrategy.cpp:
```diff
@@ -1389,6 +1528,14 @@ bool ILPInitialScheduleStage::shouldRevertScheduling(unsigned WavesAfter) {
 
+bool MemoryClauseInitialScheduleStage::shouldRevertScheduling(
+    unsigned WavesAfter) {
+  if (mayCauseSpilling(WavesAfter))
+    return true;
+
+  return false;
```
```suggestion
  return mayCauseSpilling(WavesAfter);
```

arsenm commented on llvm/lib/Target/AMDGPU/GCNSchedStrategy.cpp:
```diff
@@ -615,6 +615,138 @@ bool GCNMaxILPSchedStrategy::tryCandidate(SchedCandidate &Cand,
+      return TryCand.Reason != NoCand;
+
+    // Fall through to original instruction order.
+    if ((Zone->isTop() == (TryCand.SU->NodeNum < Cand.SU->NodeNum))) {
```
```suggestion
    if (Zone->isTop() == (TryCand.SU->NodeNum < Cand.SU->NodeNum)) {
```

Leftover parens

arsenm reviewed (approved):
lgtm with nits

ruiling committed:
Address review comments

ruiling committed:
Merge remote-tracking branch 'origin/main' into amdgpu-trycand

ruiling committed:
Update test checks after rebase
